src/RMGGrabmayrGCReader.cc

```diff
@@ -0,0 +1,161 @@
+}
+
+// Returns the next cascade for the Isotope Z, A.
+GammaCascadeLine RMGGrabmayrGCReader::GetNextEntry(G4int z, G4int a) {
+  // Find the corresponding file
+  std::pair<G4int, G4int> key = std::make_pair(z, a);
+  auto it = fCascadeFiles.find(key);
+  if (it == fCascadeFiles.end())
+    RMGLog::Out(RMGLog::fatal, "No Cascade exists for Isotope Z: " + std::to_string(z) +
+                                   " A: " + std::to_string(a) + " Exit!");
+  // read next line from file
+  std::string line;
+  do {
+    if (!std::getline(*(it->second), line)) {
+      // if end-of-file is reached, reset file and read first line
+      RMGLog::Out(RMGLog::detail, "Gamma cascade file EOF reached, re-opening the file");
+      it->second->clear();                 // clear EOF flag
+      it->second->seekg(0, std::ios::beg); // move to beginning of file
+      if (!std::getline(*(it->second), line)) {
+        RMGLog::Out(RMGLog::fatal, "Failed to read next line after re-opening the file. Exit!");
+      }
+    }
+  } while (line[0] == '%' || (line.find("version") !=
+                                 std::string::npos)); // This could be outsourced to SetStartLocation
+
+  // parse line and return as struct
+  GammaCascadeLine gamma_cascade;
+  std::istringstream iss(line);
+  iss >> gamma_cascade.en >> gamma_cascade.ex >> gamma_cascade.m >> gamma_cascade.em;
+  gamma_cascade.eg.reserve(gamma_cascade.m);
+  int eg_value;
+  for (int i = 0; i < gamma_cascade.m; i++) {
+    if (!(iss >> eg_value)) {
+      RMGLog::Out(RMGLog::fatal, "Failed to read gamma energy from file. Exit!");
+    }
+    gamma_cascade.eg.push_back(eg_value);
+  }
+  return gamma_cascade;
+}
+
+
+// Is there a better way dealing with the unique pointer?
+std::unique_ptr<std::ifstream> RMGGrabmayrGCReader::SetStartLocation(
+    std::unique_ptr<std::ifstream> file) {
+  if (!file || !file->is_open())
+    RMGLog::Out(RMGLog::fatal, "The file is not open to set start location! Exit.");
+  file->clear();                 // clear EOF flag
+  file->seekg(0, std::ios::beg); // move to beginning of file
+  // Skip Header
+  std::string line;
+  int header_length = 0;
+  do {
+    std::getline(*file, line);
+    header_length++;
+  } while (line[0] == '%' || (line.find("version") != std::string::npos));
+
+  // In case the Random start location macro is set
+  if (fGammaCascadeRandomStartLocation) {
+    int n_entries_in_file = 0;
+    // Seems excessiv to read through the entire file, there has to be a quicker way?
+    while (std::getline(*file, line)) n_entries_in_file++;
+
+    file->clear();                 // clear EOF flag
+    file->seekg(0, std::ios::beg); // move to beginning of file
+
+    int start_location = (int)(n_entries_in_file * G4UniformRand() + header_length);
+
+    RMGLog::Out(RMGLog::detail, "Random start location: ", start_location);
+    for (int j = 0; j < start_location; j++) std::getline(*file, line);
+  }
+  return std::move(file);
```

Comment:
I think you should be able to use a reference instead of a unique_ptr here, to avoid the repetitive std::move-ing around (not compile tested :-)

```suggestion
void RMGGrabmayrGCReader::SetStartLocation(std::ifstream& file) {
  if (!file.is_open())
    RMGLog::Out(RMGLog::fatal, "The file is not open to set start location! Exit.");
  file.clear();                 // clear EOF flag
  file.seekg(0, std::ios::beg); // move to beginning of file
  // Skip Header
  std::string line;
  int header_length = 0;
  do {
    std::getline(file, line);
    header_length++;
  } while (line[0] == '%' || (line.find("version") != std::string::npos));

  // In case the Random start location macro is set
  if (fGammaCascadeRandomStartLocation) {
    int n_entries_in_file = 0;
    // Seems excessiv to read through the entire file, there has to be a quicker way?
    while (std::getline(file, line)) n_entries_in_file++;

    file.clear();                 // clear EOF flag
    file.seekg(0, std::ios::beg); // move to beginning of file

    int start_location = (int)(n_entries_in_file * G4UniformRand() + header_length);

    RMGLog::Out(RMGLog::detail, "Random start location: ", start_location);
    for (int j = 0; j < start_location; j++) std::getline(file, line);
  }
```